File: robot_2_description/robot_2_description/serial_bridge.py
#!/usr/bin/env python3
import rclcpp
from rclcpp.node import Node
from std_msgs.msg import Float64MultiArray
import socket
import time

class BluetoothBridge(Node):
    def __init__(self):
        super().__init__('serial_bridge')
        
        # Địa chỉ MAC của con HC-05 của bro
        self.bd_addr = "00:23:11:A0:38:24"
        self.port = 1 # Kênh RFCOMM mặc định
        
        self.sock = None
        self.connect_to_hc05()

        # Khởi tạo giá trị ban đầu an toàn
        self.v_left = 0.0
        self.v_right = 0.0
        self.steer = 0.0

        # Đăng ký nhận dữ liệu từ các Topic đầu ra của node Ackermann C++
        self.sub_pos = self.create_subscription(Float64MultiArray, '/forward_position_controller/commands', self.pos_cb, 10)
        self.sub_vel = self.create_subscription(Float64MultiArray, '/forward_velocity_controller/commands', self.vel_cb, 10)
        
        # Vòng lặp gửi dữ liệu định kỳ 50Hz (20ms/lần) khớp hoàn toàn với update_rate của STM32
        self.timer = self.create_timer(0.02, self.send_to_stm32)

    def connect_to_hc05(self):
        try:
            self.get_logger().info(f"Đang kết nối Socket tới HC-05 [{self.bd_addr}]...")
            self.sock = socket.socket(socket.AF_BLUETOOTH, socket.SOCK_STREAM, socket.BTPROTO_RFCOMM)
            self.sock.connect((self.bd_addr, self.port))
            self.get_logger().info("🔥 KẾT NỐI BLUETOOTH THÀNH CÔNG! Thao tác xe sẵn sàng.")
        except Exception as e:
            self.get_logger().error(f"Kết nối thất bại: {e}. Đang thử lại sau 2 giây...")
            time.sleep(2)
            self.connect_to_hc05()

    def pos_cb(self, msg):
        if len(msg.data) > 0:
            self.steer = msg.data[0] # Lấy góc lái (Radian) của bánh trước từ hệ Ackermann

    def vel_cb(self, msg):
        if len(msg.data) >= 2:
            self.v_left = msg.data[0]  # Vận tốc góc bánh sau TRÁI (rad/s)
            self.v_right = msg.data[1] # Vận tốc góc bánh sau PHẢI (rad/s)

    def send_to_stm32(self):
        if self.sock is None:
            return
            
        # KHỚP LOGIC STM32: Tính vận tốc trung bình tổng của xe từ hệ vi sai ảo
        v_mean = (self.v_left + self.v_right) / 2.0
        
        # 1. Quy đổi Vận tốc góc -> Giá trị điều khiển thô (-255 đến 255) giống hệt file C++ cũ
        target_v_stm = int(v_mean * 15.0)
        target_v_stm = max(-255, min(255, target_v_stm)) # Ép dải thô giới hạn MAX_TARGET_VELOCITY

        # 2. Quy đổi Góc bẻ lái Radian -> Xung PWM Servo (1000 đến 2000, giữa là 1500)
        target_servo_stm = 1500 + int(self.steer * 400.0)
        target_servo_stm = max(1000, min(2000, target_servo_stm))

        # Đóng gói chuỗi String khít khịt với hàm sscanf(parse_buffer, "V:%d,S:%d") dưới STM32
        tx_str = f"V:{target_v_stm},S:{target_servo_stm}\n"
        
        try:
            # Bắn thẳng dữ liệu qua Bluetooth Socket
            self.sock.send(tx_str.encode('utf-8'))
        except Exception as e:
            self.get_logger().warn(f"Mất kết nối với mạch: {e}. Đang tự động kết nối lại...")
            self.connect_to_hc05()

def main(args=None):
    rclcpp.init(args=args)
    node = BluetoothBridge()
    try:
        rclcpp.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        if node.sock:
            node.sock.close()
        node.destroy_node()
        rclcpp.shutdown()

if __name__ == '__main__':
    main()

File: ackermann_control/src/serial_bridge.cpp
#include <chrono>
#include <string>
#include <memory>
#include <algorithm>
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/float64_multi_array.hpp"
#include "geometry_msgs/msg/twist.hpp"
// Thư viện serial thuần của Linux để nói chuyện với STM32
#include <fcntl.h>
#include <termios.h>
#include <unistd.h>

class SerialBridge : public rclcpp::Node
{
public:
  SerialBridge() : Node("serial_bridge")
  {
    // 1. Cấu hình cổng Serial kết nối STM32
    serial_port_ = open("/dev/ttyUSB0", O_RDWR | O_NOCTTY | O_NDELAY);
    if (serial_port_ < 0) {
      RCLCPP_ERROR(this->get_logger(), "Không thể mở cổng Serial /dev/ttyUSB0! Kiểm tra dây cáp.");
    } else {
      struct termios options;
      tcgetattr(serial_port_, &options);
      cfsetispeed(&options, B115200); // Tốc độ baudrate 115200 trùng với USART1 STM32
      cfsetospeed(&options, B115200);
      options.c_cflag |= (CLOCAL | CREAD);
      options.c_cflag &= ~PARENB;
      options.c_cflag &= ~CSTOPB;
      options.c_cflag &= ~CSIZE;
      options.c_cflag |= CS8;
      tcsetattr(serial_port_, TCSANOW, &options);
      RCLCPP_INFO(this->get_logger(), "Kết nối thành công tới STM32 tại cổng /dev/ttyUSB0");
    }

    // 2. Đăng ký nhận dữ liệu đã tính toán từ node Động học Ackermann cũ
    sub_pos_ = this->create_subscription<std_msgs::msg::Float64MultiArray>(
      "/forward_position_controller/commands", 10, std::bind(&SerialBridge::pos_cb, this, std::placeholders::_1));
    
    sub_vel_ = this->create_subscription<std_msgs::msg::Float64MultiArray>(
      "/forward_velocity_controller/commands", 10, std::bind(&SerialBridge::vel_cb, this, std::placeholders::_1));

    // Khởi tạo giá trị ban đầu
    latest_v_ = 0.0;
    latest_steer_ = 0.0;

    // Vòng lặp gửi data xuống STM32 chu kỳ 20ms (50Hz) cho mượt
    timer_ = this->create_wall_timer(std::chrono::milliseconds(20), std::bind(&SerialBridge::send_to_stm32, this));
  }

  ~SerialBridge() {
    if (serial_port_ >= 0) close(serial_port_);
  }

private:
  void pos_cb(const std_msgs::msg::Float64MultiArray::SharedPtr msg) {
    if(!msg->data.empty()) {
      latest_steer_ = msg->data[0]; // Lấy góc lái (Radian)
    }
  }

  void vel_cb(const std_msgs::msg::Float64MultiArray::SharedPtr msg) {
    if(!msg->data.empty()) {
      latest_v_ = msg->data[0]; // Lấy vận tốc góc bánh xe (rad/s)
    }
  }

  void send_to_stm32()
  {
    if (serial_port_ < 0) return;

    // --- QUY ĐỔI ĐƠN VỊ TỪ ROS 2 SANG STM32 ---
    
    // 1. Quy đổi Vận tốc góc (rad/s) -> Giá trị điều khiển từ -255 đến 255
    // Hệ số 15.0 này bro cần tinh chỉnh lại tùy thuộc vào lốp và áp sạc lốp tế tế
    int target_v_stm = static_cast<int>(latest_v_ * 15.0); 
    target_v_stm = std::clamp(target_v_stm, -255, 255);

    // 2. Quy đổi Góc bẻ lái Radian -> Xung PWM Servo (1000 đến 2000, giữa là 1500)
    // Hệ số bẻ lái góc: 1 Radian ~ 57.3 độ. Tùy thuộc vào cơ cấu trục bánh thực tế
    int target_servo_stm = 1500 + static_cast<int>(latest_steer_ * 400.0);
    target_servo_stm = std::clamp(target_servo_stm, 1000, 2000);

    // Đóng gói thành chuỗi String đúng định dạng STM32 đang đợi đọc: "V:%d,S:%d\n"
    std::string tx_str = "V:" + std::to_string(target_v_stm) + ",S:" + std::to_string(target_servo_stm) + "\n";

    // Bắn thẳng xuống UART
    write(serial_port_, tx_str.c_str(), tx_str.length());
  }

  int serial_port_;
  rclcpp::Subscription<std_msgs::msg::Float64MultiArray>::SharedPtr sub_pos_;
  rclcpp::Subscription<std_msgs::msg::Float64MultiArray>::SharedPtr sub_vel_;
  rclcpp::TimerBase::SharedPtr timer_;

  double latest_v_;
  double latest_steer_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<SerialBridge>());
  rclcpp::shutdown();
  return 0;
}